.circleci/config.yml

```diff
@@ -75,106 +66,327 @@ commands:
             esac
             echo "Setting deployment ENVIRONMENT to $DEPLOY_ENV"
             echo "export ENVIRONMENT=$DEPLOY_ENV" >> $BASH_ENV
-            echo 'export NG_BUILD_OPTS=$NG_BUILD_OPTS' >> $BASH_ENV
             source $BASH_ENV
+
+  deploy-stored-build:
+    description: Deploy archived build to GAE
+    parameters:
+      study_key:
+        type: string
+      study_guid:
+        type: string
+    steps:
+      - checkout:
+          path: ~/repo
+      - setup-shared-env:
+          study_key: << parameters.study_key >>
       - run:
-          name: <<parameters.study_key>> show environment variables
-          command: env
+          name: Retrieve and expand build archive for << parameters.study_key >> with git SHA << pipeline.git.revision >>
+          command: |
+            set -u
+            # we are getting our builds from the one bucket
+            ./readvault.sh secret/pepper/dev/v1/conf .data.gcp.serviceKey | gcloud auth activate-service-account --key-file=-
+            TAR_FILE_URL_PATTERN="gs://${BUILDS_BUCKET}/${ANGULAR_PROJECT_NAME}/${ANGULAR_PROJECT_NAME}_*_${SHORT_GIT_SHA}.tar.gz"
+            TAR_FILE_URL=`gsutil ls $TAR_FILE_URL_PATTERN  | sort -r | head -1`
+
```

Comment:
Here we look for through the  build archive (if multiple builds of the same SHA pickup the newest one).

Reply:
Maybe we have already explored it and that I missed the discussion: will it be possible to use CircleCI's Artifacts storage instead of buckets?

Reply:
It's a good question, Yufeng.
I had assumed that we would use the CircleCi artifact capabilities for handling Angular builds and back-end jars.
On a close reading of the docs and blog posts, though it turns out that their "artifact" support is extremely limited: their configuration API allow for storing the artifacts and not reading them. Furthermore, they warn in fine print:They are available for download from a web page on the circleci website but:

Reply:
Those are good considerations. I have not looked at their docs too closely, but it does seems like you can "read" the artifacts by calling their API with `curl`. And you can get the latest artifact too! This seems comparable to downloading from our bucket.

The fine print does sound concerning, but I don't think we need these build artifacts to be long-lived since most of the time we're deploying them within a sprint or two? Unless we _do in fact_ want to keep build artifacts around... If not and we keep with the buckets, maybe we should have a process to occasionally delete old artifacts from buckets (especially dev/test builds since those can accumulate quickly). I just hope we don't end up hand-rolling our own artifact management system.

Reply:
You can indeed use `curl` to download files and get the latest, but for some reason they don't provide a way to list them or a configuration option to retrieve a build to use on a deployment that is the counterpart to saving the artifact (something that I found very disconcerting).
The fact that they warn you about not depending on artifacts long term without telling you how long you can indeed count on them (a day, a month, a year?) being there is troubling.
My meta-take away was that the feature was half-cooked and treated as an afterthought.
Replacing it with our own bucket seemed safer and just as easy for what we needed at this time.
Maybe they'll improve the feature or we could jump on Artifactory if our needs grew in complexity. Either ought to be pretty easy to plug into place.
Cleanup can eventually be a concern. We could look at removing builds that are older than month that are not currently deployed or something like that?

Reply:
Sounds good. I just found out that Google Buckets [has nifty features to automate object deletion](https://cloud.google.com/storage/docs/lifecycle) so that's something we can explore.

Reply:
I'll create a ticket to add cleanup. I think our needs will be a little more complicated than just letting GCP remove old files. We probably will want to query the GCP to find out  what is currently running there and then remove builds that have not run in a while (where "a while" is TBD).